src/ucp/proto/proto_common.c

```diff
@@ -392,163 +426,249 @@ ucp_proto_common_get_reg_cost(const ucp_proto_common_init_params_t *params,
-    range->max_length = SIZE_MAX;
-    range->perf       = ucs_linear_func_make(0, 0);
+    return UCS_OK;
+}
 
-    if (ucs_test_all_flags(params->flags, UCP_PROTO_COMMON_INIT_FLAG_SEND_ZCOPY |
-                                          UCP_PROTO_COMMON_INIT_FLAG_RECV_ZCOPY)) {
-        ucs_linear_func_add_inplace(&range->perf, uct_time);
-    } else {
-        piped_send_cost = ucs_linear_func_make(0, ucs_max(pack_time.m, uct_time.m));
-        piped_size      = ucs_linear_func_make(-1.0 * frag_size, 1);
+/*
+ * Calculate the performance pipelining an operation with performance 'perf1'
+ * with an operation with performance 'perf2' by using fragments with size
+ * 'frag_size' bytes.
+*/
+static ucs_linear_func_t ucp_proto_common_ppln_perf(ucs_linear_func_t perf1,
+                                                    ucs_linear_func_t perf2,
+                                                    double frag_size)
+{
+    double adjusted_frag_size = ucs_max(frag_size, 1.0);
+    double max_m              = ucs_max(perf1.m, perf2.m);
+    ucs_linear_func_t result;
 
-        /* Copy first fragment */
-        if (!(params->flags & UCP_PROTO_COMMON_INIT_FLAG_SEND_ZCOPY)) {
-            ucs_linear_func_add_value_at(&range->perf, pack_time, frag_size);
-        }
+    /* Pipeline overhead is maximal of both protocols' overheads */
+    result.c = ucs_max(perf1.c, perf2.c);
 
-        /* Reach the point where we can start sending the last fragment */
-        ucs_linear_func_add_inplace(&range->perf,
-                                    ucs_linear_func_compose(piped_send_cost,
-                                                            piped_size));
+    /*
+     * Pipeline cost per byte is maximal of both protocols' cost per byte.
+     * If the overhead per fragment is higher than the cost per byte, we take
+     * the overhead per fragment divided by the fragment size.
+     *
+     * NOTE: We don't add the fragment overhead to the cost per byte, because we
+     * assume multiple fragments are sent, so the overhead per fragment can be
+     * hidden by the cost per byte of the same operation.
+     */
+    result.m = ucs_max(max_m, result.c / adjusted_frag_size);
 
-        /* Send last fragment */
-        ucs_linear_func_add_value_at(&range->perf, uct_time, frag_size);
-    }
+    ucs_assert((result.m >= perf1.m) && (result.m >= perf2.m) &&
+               (result.c >= perf1.c) && (result.c >= perf2.c));
+    return result;
+}
 
-    /* Receive last fragment */
-    ucs_linear_func_add_value_at(&range->perf, recv_time, frag_size);
+static ucs_linear_func_t
+ucp_proto_common_ppln3_perf(ucs_linear_func_t perf1, ucs_linear_func_t perf2,
+                            ucs_linear_func_t perf3, double frag_size)
+{
+    return ucp_proto_common_ppln_perf(ucp_proto_common_ppln_perf(perf1, perf2,
+                                                                 frag_size),
+                                      perf3, frag_size);
```

Comment:
is it for 3 stage pipeline or there is some other use case?

Reply:
it's used by ucp_proto_common_init_caps

Reply:
a, yes. it's more like a stages of a protocol

Reply:
yes, it's like the paper you send which models performance as send_overhead,transfer,recv_overhead